src/index.ts

```diff
@@ -1 +1,171 @@
+      const variableRegex = /\$\{([a-zA-Z_][a-zA-Z0-9_]*)\}/g;
+      const params: string[] = [];
+      while (true) {
+        const match = variableRegex.exec(jsonObj);
+        if (match === null) break;
+        if (!params.includes(match[1])) params.push(match[1]);
+      }
+      return new FunctionType(params);
+    } else if (typeof jsonObj === 'object' && !Array.isArray(jsonObj)) {
+      const map = new Map<string, BaseType>();
+      for (const [key, value] of Object.entries(jsonObj)) {
+        const valueTreeType = jsonObjToTypeObj(value);
+        map.set(key, valueTreeType);
+      }
+      return new ObjectType(map);
+    }
+    throw new Error('Error: JSONObject is neigher string nor object');
+  };
+
+  const jsonObj = JSON.parse(fs.readFileSync(langFilepath, { encoding: 'utf-8', flag: 'r' }));
+  return jsonObjToTypeObj(jsonObj);
+}
+
+function validate(langFilepath: string, typeObj: BaseType): void {
+  const equals = (ourTypeObj: BaseType, theirTypeObj: BaseType): boolean => {
+    if (ourTypeObj instanceof FunctionType && theirTypeObj instanceof FunctionType) {
+      if (ourTypeObj.params.length !== theirTypeObj.params.length) return false;
+      const sortedOurParams = ourTypeObj.params.slice().sort();
+      const sortedTheirParams = theirTypeObj.params.slice().sort();
+      for (let i = 0; i < ourTypeObj.params.length; i++) {
+        if (sortedOurParams[i] !== sortedTheirParams[i]) return false;
+      }
+      return true;
+    } else if (ourTypeObj instanceof ObjectType && theirTypeObj instanceof ObjectType) {
+      if (ourTypeObj.map.size !== theirTypeObj.map.size) return false;
+      for (const [key, value] of ourTypeObj.map) {
+        const child = theirTypeObj.map.get(key);
+        if (child === undefined || !equals(value, child)) return false;
+      }
+      return true;
+    }
+    return false;
+  };
+
+  if (!equals(typeObj, convert(langFilepath))) throw new Error('Error: validation failed');
+}
+
+function gen(langFilepaths: string[], typeObj: BaseType, defaultLang: string): string {
+  const langs = langFilepaths.map((langFilepath) => path.parse(langFilepath).name);
+  if (!langs.includes(defaultLang)) {
+    throw new Error('Error: cannot find default-lang file');
+  }
+
+  const varCurrentLang = 'currentLang';
+  let codeString = '';
+
+  const jsonObjToCodeString = (jsonObj: any): string => {
+    if (typeof jsonObj === 'string') {
+      return `"${jsonObj}"`;
+    } else if (typeof jsonObj === 'object' && !Array.isArray(jsonObj)) {
+      let members = '';
+      for (const [key, value] of Object.entries(jsonObj)) {
+        const valueCodeString = jsonObjToCodeString(value);
+        members += `${key}: ${valueCodeString}, `;
+      }
+      return `{ ${members} }`;
+    }
+    throw new Error('Error: JSONObject is neigher string nor object');
+  };
+
+  for (const langFilepath of langFilepaths) {
+    const fileText = fs.readFileSync(langFilepath, { encoding: 'utf-8', flag: 'r' });
```

Comment:
`flag: 'r'` should be omitted because we can easily understand `readFileSync` reads (and doesn't writes) a file .